Sources/ArgumentParser/Completions/BashCompletionsGenerator.swift

```diff
@@ -9,257 +9,397 @@
+            case "${subcommand}" in
+            \(subcommands.map { $0._commandName }.joined(separator: "|")))
+                # Offer subcommand argument completions
+                "\(functionName)_${subcommand}"
+                ;;
+            *)
+                # Offer subcommand completions
+                COMPREPLY+=($(compgen -W '\(
+                  subcommands.map { $0._commandName.shellEscapeForSingleQuotedString() }.joined(separator: " ")
+                )' -- "${cur}"))
+                ;;
+            esac
 
-    return result
-      + subcommands
-      .map { generateCompletionFunction(commands + [$0]) }
-      .joined()
-  }
+        """
+    }
 
-  /// Returns the option and flag names that can be top-level completions.
-  fileprivate static func generateArgumentWords(
-    _ commands: [ParsableCommand.Type]
-  ) -> [String] {
-    commands
-      .argumentsForHelp(visibility: .default)
-      .flatMap { $0.bashCompletionWords() }
-  }
+    if result.isEmpty {
+      result = "    :\n"
+    }
 
-  /// Returns additional top-level completions from positional arguments.
-  ///
-  /// These consist of completions that are defined as `.list` or `.custom`.
-  fileprivate static func generateArgumentCompletions(
-    _ commands: [ParsableCommand.Type]
-  ) -> [String] {
-    guard let type = commands.last else { return [] }
-    return ArgumentSet(type, visibility: .default, parent: nil)
-      .compactMap { arg -> String? in
-        guard arg.isPositional else { return nil }
-
-        switch arg.completion.kind {
-        case .default, .file, .directory:
-          return nil
-        case .list(let list):
-          return list.joined(separator: " ")
-        case .shellCommand(let command):
-          return "$(\(command))"
-        case .custom:
-          return """
-            $("${COMP_WORDS[0]}" \(arg.customCompletionCall(commands)) "${COMP_WORDS[@]}")
-            """
-        }
+    return """
+      \(functionName)() {
+      \(result)\
       }
-  }
 
-  /// Returns the case-matching statements for supplying completions after an option or flag.
-  fileprivate static func generateOptionHandlers(
-    _ commands: [ParsableCommand.Type]
-  ) -> String {
-    guard let type = commands.last else { return "" }
-    return ArgumentSet(type, visibility: .default, parent: nil)
-      .compactMap { arg -> String? in
-        let words = arg.bashCompletionWords()
-        if words.isEmpty { return nil }
-
-        // Flags don't take a value, so we don't provide follow-on completions.
-        if arg.isNullary { return nil }
+      \(subcommands.map { (self + [$0]).completionFunctions }.joined())
+      """
+  }
 
-        return """
-          \(arg.bashCompletionWords().joined(separator: "|")))
-          \(arg.bashValueCompletion(commands).indentingEachLine(by: 4))
-              return
-          ;;
-          """
+  /// Returns flags for the last command of the given array.
+  private var flags: [String] {
```

Comment:
Thoughts on naming this something like `flagCompletions` instead? Reading the source above it looked like it would be the same type as `positionalArguments` (an array of `ArgumentDefinition`). Same thoughts for `options`, below.